File: sdks/ios/Sources/Likerts/AdvancedQuestions.swift
import Foundation

public enum AdvancedQuestions {
  public static func move(_ order:[String], id:String, offset:Int)->[String] { guard let from=order.firstIndex(of:id) else{return order};let to=max(0,min(order.count-1,from+offset));var next=order;next.remove(at:from);next.insert(id,at:to);return next }
  public static func setMatrix(_ answer:[String:[String]],row:String,column:String,multiple:Bool)->[String:[String]] { var next=answer;if !multiple{next[row]=[column];return next};var selected=next[row] ?? [];if let index=selected.firstIndex(of:column){selected.remove(at:index)}else{selected.append(column)};if selected.isEmpty{next.removeValue(forKey:row)}else{next[row]=selected};return next }
  public static func remaining(total:Int,answer:[String:Int])->Int { total-answer.values.reduce(0,+) }
  public static func validRanking(_ answer:[String],options:[String])->Bool { answer.count==options.count && Set(answer).count==answer.count && answer.allSatisfy(options.contains) }
  public static func validMatrix(_ answer:[String:[String]],rows:[String],columns:[String],multiple:Bool,required:Bool)->Bool { !answer.isEmpty && (!required || answer.count==rows.count) && answer.allSatisfy { row,selected in rows.contains(row) && (multiple ? !selected.isEmpty : selected.count==1) && Set(selected).count==selected.count && selected.allSatisfy(columns.contains) } }
  public static func validAllocation(_ answer:[String:Int],items:[String],total:Int)->Bool { answer.count==items.count && answer.keys.allSatisfy(items.contains) && answer.values.allSatisfy{$0>=0&&$0<=total} && remaining(total:total,answer:answer)==0 }
}

File: sdks/android/src/main/kotlin/com/likerts/sdk/AdvancedQuestions.kt
package com.likerts.sdk

object AdvancedQuestions {
    fun move(order:List<String>,id:String,offset:Int):List<String>{val from=order.indexOf(id);if(from<0)return order;val to=(from+offset).coerceIn(0,order.lastIndex);return order.toMutableList().apply{removeAt(from);add(to,id)}}
    fun setMatrix(answer:Map<String,List<String>>,row:String,column:String,multiple:Boolean):Map<String,List<String>>{val next=answer.toMutableMap();if(!multiple){next[row]=listOf(column);return next};val selected=(next[row]?:emptyList()).toMutableList();if(column in selected)selected.remove(column)else selected+=column;if(selected.isEmpty())next.remove(row)else next[row]=selected;return next}
    fun remaining(total:Int,answer:Map<String,Int>)=total-answer.values.sum()
    fun validRanking(answer:List<String>,options:List<String>)=answer.size==options.size&&answer.toSet().size==answer.size&&answer.all{it in options}
    fun validMatrix(answer:Map<String,List<String>>,rows:List<String>,columns:List<String>,multiple:Boolean,required:Boolean)=answer.isNotEmpty()&&(!required||answer.size==rows.size)&&answer.all{(row,selected)->row in rows&&(if(multiple)selected.isNotEmpty()else selected.size==1)&&selected.toSet().size==selected.size&&selected.all{it in columns}}
    fun validAllocation(answer:Map<String,Int>,items:List<String>,total:Int)=answer.size==items.size&&answer.keys.all{it in items}&&answer.values.all{it in 0..total}&&remaining(total,answer)==0
}
